# Assignment Four

#### Load Data and Packages

In [1]:
import pandas as pd
from textblob import TextBlob
import string

In [2]:
DC = pd.read_csv(r"C:\Users\datre\OneDrive\Documents\Graduate School\Winter '20\DailyComments.csv")

In [3]:
DC

Day of Week                                        comments
0      Monday                             Hello, how are you?
1     Tuesday                            Today is a good day!
2   Wednesday  It's my birthday so it's a really special day!
3    Thursday       Today is neither a good day or a bad day!
4      Friday                           I'm having a bad day.
5    Saturday       There' s nothing special happening today.
6      Sunday                      Today is a SUPER good day!

In [4]:
comment = DC["comments"]
comment

0                               Hello, how are you?
1                              Today is a good day!
2    It's my birthday so it's a really special day!
3         Today is neither a good day or a bad day!
4                             I'm having a bad day.
5         There' s nothing special happening today.
6                        Today is a SUPER good day!
Name: comments, dtype: object

#### Sentiment Analysis

I will be using the TextBlob sentiment analysis for this assignment. The sentiment analysis of the TextBlob function displays two analyses of the statements. The first is polarity, meaning the positivity or negativity of the statement. This measurement ranges from -1, which is very negative, to 1, which is very positive. This means a neutral statement lies at 0. The second measurement displayed is the subjectivity of the statement, meaning the persons emotion, judgement, or opinion in the sentence. This measurement is from 0 to 1 with 1 being the highest amount of subjectivity and 0 being no subjectivity. 

In [5]:
def remove_punctuations(text):
    for punctuation in string.punctuation:
        text = text.replace(punctuation, '')
    return text

In [6]:
DC["nopunc"] = DC['comments'].apply(remove_punctuations)
DC["clean"] = DC["nopunc"].str.lower()
DC

Day of Week                                        comments  \
0      Monday                             Hello, how are you?   
1     Tuesday                            Today is a good day!   
2   Wednesday  It's my birthday so it's a really special day!   
3    Thursday       Today is neither a good day or a bad day!   
4      Friday                           I'm having a bad day.   
5    Saturday       There' s nothing special happening today.   
6      Sunday                      Today is a SUPER good day!   

                                        nopunc  \
0                            Hello how are you   
1                          Today is a good day   
2  Its my birthday so its a really special day   
3     Today is neither a good day or a bad day   
4                          Im having a bad day   
5      There s nothing special happening today   
6                    Today is a SUPER good day   

                                         clean  
0                            hello how are you  
1                          today is a good day  
2  its my birthday so its a really special day  
3     today is neither a good day or a bad day  
4                          im having a bad day  
5      there s nothing special happening today  
6                    today is a super good day

In [7]:
no_punc = DC["nopunc"]
clean_com = DC["clean"]

In [8]:
for comments in comment:
    print(comments)
    analysis = TextBlob(comments)
    print(analysis.sentiment)

Hello, how are you?
Sentiment(polarity=0.0, subjectivity=0.0)
Today is a good day!
Sentiment(polarity=0.875, subjectivity=0.6000000000000001)
It's my birthday so it's a really special day!
Sentiment(polarity=0.44642857142857145, subjectivity=0.5714285714285714)
Today is neither a good day or a bad day!
Sentiment(polarity=-0.08749999999999991, subjectivity=0.6333333333333333)
I'm having a bad day.
Sentiment(polarity=-0.6999999999999998, subjectivity=0.6666666666666666)
There' s nothing special happening today.
Sentiment(polarity=0.35714285714285715, subjectivity=0.5714285714285714)
Today is a SUPER good day!
Sentiment(polarity=0.6041666666666666, subjectivity=0.6333333333333333)


Monday: The sentence does not display any emotion, it is merely a question. For this reason the analysis resutled in 0 polarity and 0 subjectivity. If there was an additional word, say an endeering term if talking to someone you care about, this result may be different.

Tuesday: This sentence is labeled very positive beause of the word good as well as the exclamation point (more below on that). It also scored high on the sibjectivity, which makes sense because one person can think today is good while another might have a different opinion based on events going on or their likes and dislikes.

Wednesday: This scored lower on the polarity than I expected. However, it does make sense since there are no obvious "good" words such as good, great, awesome. Special could be a more positive word that is helping with this score. 

Thursday: The sentence comes off as neutral to us, but since the word neither is involved in the sentence, I feel the scheme places it in the slightly negative category.

Friday: This is the most negative sentence in the group. With no potisitve punctuation and a negative word "bad", this sentence is sent to the negative pile.

Saturday: Again, this is a fairly neutral sentence according to humans. When the machine looks at it, the word special is similar to the preious word neither but on the other extreme. It is a slightly positive word. Because of this, this farily neutral sentence sways over to the positive side.

Sunday: The last sentence is a pretty positive sentence. By using "super", "good", and an exclamation point, this sentence earns very positive marks.

#### Analysis with cleaner data

In [9]:
for nopunc in no_punc:
    print(nopunc)
    analysis = TextBlob(nopunc)
    print(analysis.sentiment)

Hello how are you
Sentiment(polarity=0.0, subjectivity=0.0)
Today is a good day
Sentiment(polarity=0.7, subjectivity=0.6000000000000001)
Its my birthday so its a really special day
Sentiment(polarity=0.35714285714285715, subjectivity=0.5714285714285714)
Today is neither a good day or a bad day
Sentiment(polarity=5.551115123125783e-17, subjectivity=0.6333333333333333)
Im having a bad day
Sentiment(polarity=-0.6999999999999998, subjectivity=0.6666666666666666)
There s nothing special happening today
Sentiment(polarity=0.35714285714285715, subjectivity=0.5714285714285714)
Today is a SUPER good day
Sentiment(polarity=0.5166666666666666, subjectivity=0.6333333333333333)


In [10]:
for clean in clean_com:
    print(clean)
    analysis = TextBlob(clean)
    print(analysis.sentiment)

hello how are you
Sentiment(polarity=0.0, subjectivity=0.0)
today is a good day
Sentiment(polarity=0.7, subjectivity=0.6000000000000001)
its my birthday so its a really special day
Sentiment(polarity=0.35714285714285715, subjectivity=0.5714285714285714)
today is neither a good day or a bad day
Sentiment(polarity=5.551115123125783e-17, subjectivity=0.6333333333333333)
im having a bad day
Sentiment(polarity=-0.6999999999999998, subjectivity=0.6666666666666666)
there s nothing special happening today
Sentiment(polarity=0.35714285714285715, subjectivity=0.5714285714285714)
today is a super good day
Sentiment(polarity=0.5166666666666666, subjectivity=0.6333333333333333)


I wanted to analyze the same set with the same function but changing the cleanliness of the data. I found by removing punctuation, TextBlob will change the polarity of the sentence (subjectivity is the same). All of the entries with exclamation points scored more negative than the original when the exclamation points were removed. The first, "Today is a good day!" went from 0.875 to 0.7 in polarity score.  

#### AirBnb Reviews

In [12]:
reviews = pd.read_csv(r"C:\Users\datre\OneDrive\Documents\Graduate School\Winter '20\reviews.csv")
reviews.head()

listing_id        id        date  reviewer_id reviewer_name  \
0        3781  37776825   7/10/2015     36059247          Greg   
1        3781  41842494    8/9/2015     10459388           Tai   
2        3781  45282151    9/1/2015     12264652        Damien   
3        3781  49022647   9/30/2015     41426327          Mike   
4        3781  52503327  10/30/2015     15151513          Ivan   

                                            comments  
0  The apartment was as advertised and Frank was ...  
1  It was a pleasure to stay at Frank's place. Th...  
2  The apartment description is entirely faithful...  
3  Thoroughly enjoyed my time at Frank's home. Ha...  
4  Great value for the money! This location has e...

In [15]:
del reviews["listing_id"]
del reviews["id"]
del reviews["date"]
del reviews["reviewer_id"]

In [17]:
rev_samp = reviews.sample(n=100)
rev_samp.head()

reviewer_name                                           comments
149337       Ricarda  We had a very pleasant stay at Boaz‘s. Would r...
157231       Mariana  Fred´s place exceeded expectations in every wa...
73538          Jason  We loved our stay in the Northend, the locatio...
188608        Meghan                   Great location and trendy place!
78956          Logan  Thanks to Yonatan for a great place to stay wh...

In [26]:
pd.set_option("display.max_rows", None)

In [27]:
pol = lambda x: TextBlob(x).sentiment.polarity
sub = lambda x: TextBlob(x).sentiment.subjectivity

rev_samp["polarity"] = rev_samp["comments"].apply(pol)
rev_samp["subjectivity"] = rev_samp["comments"].apply(sub)
rev_samp

reviewer_name                                           comments  \
149337          Ricarda  We had a very pleasant stay at Boaz‘s. Would r...   
157231          Mariana  Fred´s place exceeded expectations in every wa...   
73538             Jason  We loved our stay in the Northend, the locatio...   
188608           Meghan                   Great location and trendy place!   
78956             Logan  Thanks to Yonatan for a great place to stay wh...   
92374             Clive  A comfy clean and convenient place to stay wit...   
189418            Yanxi               Two beautiful cats, be nice to them.   
89537           Douglas                                                  B   
167117            Henry  If someone were to give an example of THE perf...   
125294           Auston  Amazing, clean and beautiful space the ac was ...   
182445           Alexia  The host were amazing and very helpful. I woul...   
81352               Tim  Liesl's apartment is in a great location. It i...   
3594           Hongjing  Michelle is very nice and kind. She makes you ...   
143523            Paula  Anthony's North End Studio is in a fabulous lo...   
41970             Haley  I enjoyed my stay form check in to check out. ...   
115586             Shan  It's a loverly apartment with spacious and cle...   
74757           Lincoln  Great stay. Basic space, but fit the needs of ...   
90391          Jayamary  Erika is very accommodating and nice. They put...   
45453            Destry  Jonathan thinks of everything to help you feel...   
62171              Anne  Logement confortable et bien équipé avec une t...   
138130          Miodrag  NIce small studio at perfect location. The bac...   
158413            Holly  Great location, communication and easy check i...   
158835        Elizabeth  This unit is in an ideal location to Boston, b...   
9693             Steven  Kelly was great and super responsive. Everythi...   
180047          Rogerio  Everything perfect. I really recommend this Ka...   
68157           Michael                                          Loved it.   
56566            Henrik  Julie is an excellent host and very attentive ...   
112734             Jena   Loved the space in the apartment and the loca...   
15325          Fabrizio  I was very happy in my stay.\r\n- the location...   
82343           Heather  It was so nice to stay in Marcia's home  and I...   
73432            Nathan  Frank is a wonderful host and the place is lov...   
18841               Dan  Izzy is a gracious host and great to talk with...   
130805           Sholto                           Great location and value   
6459             Mengya  Just arrived at home. The experience of stayin...   
71571            Nicola  Fabulous apartment with an amazing view of Bos...   
165320           Thomas             Great place for exploring Boston from!   
29767              Lisa  Liz's place is beautiful and updated. It was c...   
56133             Anand  We checked in around Christmas eve. The room w...   
54523             Karen  Jay's place was very clean and exactly as pict...   
27531              Beau  The house and room were clean and easily acces...   
92867           Jessica  Sarah and Josh are exceptional hosts - they ar...   
58548            Andrew  Our stay could not have been better! Penny was...   
103965            Susie  Danny's place is very comfortable and is situa...   
42944          Jennifer  Wonderful experience! Well thought and decorat...   
18202           Douglas  Great location. Just walking distance from MBT...   
74916            George  This place was perfect. Located in a safe Lati...   
25651             Chuck  Great location for the Convention Center. Try ...   
66064             Marta  Just had a wonderful stay in the apartment. De...   
136578            Elena  Emelyn's place is very nice: she keeps it impe...   
97229             Karen  Beautiful and quiet unit in a great location. ...   
7858            Lourdes  We stay

In [25]:
rev_samp.describe()

polarity  subjectivity
count  100.000000    100.000000
mean     0.456234      0.627968
std      0.250247      0.190845
min     -0.040278      0.000000
25%      0.306052      0.554669
50%      0.414687      0.639833
75%      0.616667      0.750000
max      1.000000      1.000000

This dataset was downloaded from DataWorld at, https://data.world/antonio-alves/airbnb-reviews/workspace/file?filename=reviews.csv

With this set I removed most of the columns, keeping the names of the reviewers and the comments. From here, I applied the TextBlob analysis to each entry, adding a column for polarity and subjectivity to create a cleaner description of the results than the original problem above. I decided to take a sample of 100 from this set so that the dataframe was not terribly long. I found from this that the TextBlob funtion cannot decipher languages not using the Roman Alphabet. In addition, the overall mood of the comments was fairly positive, with a polarity mean of 0.46. the subjectivity was also high, which makes sense since these are reviews of airbnb locations. 